crates/turborepo-scm/src/lib.rs

```diff
@@ -40,43 +40,54 @@ impl Error {
     pub(crate) fn git_error(s: impl Into<String>) -> Self {
         Error::Git(s.into(), Backtrace::capture())
     }
-
-    fn from_git_exit_code(cmd: &str, path: &AbsoluteSystemPath, exit_code: Option<i32>) -> Error {
-        let s = format!(
-            "'{}' in {} exited with status code {}",
-            cmd,
-            path.to_string(),
-            exit_code
-                .map(|code| code.to_string())
-                .unwrap_or("unknown".to_string())
-        );
-        Error::Git(s, Backtrace::capture())
-    }
 }
 
-pub(crate) fn read_git_error<R: Read>(stderr: &mut R) -> Option<Error> {
+fn read_git_error_to_string<R: Read>(stderr: &mut R) -> Option<String> {
     let mut buf = String::new();
     let bytes_read = stderr.read_to_string(&mut buf).ok()?;
     if bytes_read > 0 {
         // something failed with git, report that error
-        Some(Error::git_error(buf))
+        Some(buf)
     } else {
         None
     }
 }
 
-pub(crate) fn wait_for_success<R: Read>(
+pub(crate) fn wait_for_success<R: Read, T>(
     mut child: Child,
     stderr: &mut R,
     command: &str,
     root_path: impl AsRef<AbsoluteSystemPath>,
-) -> Result<(), Error> {
+    parse_result: Result<T, Error>,
```

Comment:
I'm not in love with this API, cause it feels like this function is doing two things now: waiting for success and handling parse errors. Could we factor out the stderr reading and git error construction into a different function, then have the parse error case and the process error case both call it?

Reply:
That was the one of the problems this PR is attempting to solve: trying to read `stderr` too early due to a parse error, via the factored out `read_git_error` caused a deadlock because we had not yet called `.wait()`. So now they are intentionally coupled. A variety of things can go wrong, `wait_for_success` tries to provide as much information in an error as is available.